### PR TITLE
implement (optional) SQLite support

### PR DESCRIPTION
**Pull request by steffengy**
This implements optional SQLite support, which can be enabled using the `sqlite`
build tag.

This is a quite early version, just sharing it because I needed it anyways.

references #83

### CODE REVIEW

**Comment by aarondl** [comment]
Thank you very much for this! I'll attempt some code review soon in the future.

**Comment by millerlogic** [comment]
Thanks for this, I've been hoping for sqlite support.

While giving it a try I ran into an issue with a composite primary key.  When getting the table_info, the pk column should be int instead of bool. It appears to indicate the order.

```sql
sqlite> .head on
sqlite> pragma table_info('mytest');
cid|name|type|notnull|dflt_value|pk
0|a_id|INTEGER|1||1
1|b_id|INTEGER|1||2
```

```
Error: sql: Scan error on column index 5: sql/driver: couldn't convert 2 into type bool
unable to scan for table mytest
github.com/vattle/sqlboiler/bdb/drivers.(*SQLiteDriver).tableInfo
        /home/dev/go/src/github.com/vattle/sqlboiler/bdb/drivers/sqlite.go:136
```

**Comment by steffengy** [comment]
@millerlogic 
Yep, fixed.
```
The "pk" column in the result set is zero for columns that are not part of the primary key
 and is the index of the column in the primary key for columns that are part of the primary key.
```

**Comment by aarondl** [comment]
Hey @steffengy, just to update you on this. We're actively working on a change that will abstract drivers into separate binaries and plan on getting this in after that. Hold on just a little while and we'll get to this! Thanks for your patience.

**Comment by millerlogic** [comment]
Found another thing, the column types are case sensitive. If I create a not null datetime column, the sqlboiler type is string, if I use DATETIME it comes out as time.Time

**Comment by millerlogic** [comment]
It appears that autoincrement isn't working. I try to insert two records and it complains that the UNIQUE constraint is violated on the primary key, it looks like it's trying to insert both with id 0. [Looked up](https://stackoverflow.com/questions/18694393/how-could-you-get-if-a-table-is-autoincrement-or-not-from-the-metadata-of-an-sql) how to detect if the primary key is autoincrement, seems the only way is to look for the word autoincrement for the table's sql in sqlite_master.

**Comment by steffengy** [comment]
I'm not quite sure how to handle AUTOINCREMENT at this point,
The mysql driver simply seems to use `auto_increment` as default value when a column
is defined as `auto increment`, where I can't quite figure out how that works.

Does Mysql interpret `auto_increment` as the next value or is it translated somewhere in sqlboiler?

**Comment by aarondl** [comment]
@steffengy `auto_increment` is a mysql concept. All databases we're using today have the ability to do auto increment as a column. So when you omit the value for it it will insert a special value.

SQLBoiler handles this like so:
- Whenever a pkey column that has a default value we assume it's an auto increment column and that the database will provide a value.
- The `database/sql` library provides https://golang.org/pkg/database/sql/#Result which has `LastInsertId` which in the mysql driver sets here https://github.com/go-sql-driver/mysql/blob/ae5a66df4d33ba404d639bd0fc9189cd5ee79bcf/packets.go#L584

The sqlite drivers should support this: https://github.com/mattn/go-sqlite3/blob/46e826d22aebbfda29523e86e3114c2beb4dab43/sqlite3.go#L833

So basically if your sqlite driver says that it uses lastinsertid (which it does: https://github.com/vattle/sqlboiler/pull/138/files#diff-c880f21b86d224c850e56d2d0ce28501R57) then it should just work using the templates here: https://github.com/vattle/sqlboiler/blob/master/templates/15_insert.tpl#L82

It's possible there's a bug. I can help debug if necessary.

**Comment by aarondl** [comment]
Reading my own message one of the problems could be that the default value detection in the sqlite driver may not be working correctly?

**Comment by steffengy** [comment]
@aarondl 
Thanks that clears things up.

So to detect autoincrement for sqlite I'll have to return a similar default
value as sqlboiler's mysql driver does. 
Detection of that is a bit icky, will probably be very similar
to https://github.com/vattle/sqlboiler/pull/138#issuecomment-301974304.

**Comment by aarondl** [comment]
It should "just work" if the following is bug-free:

Default value checks:
https://github.com/vattle/sqlboiler/pull/138/files#diff-c880f21b86d224c850e56d2d0ce28501R213
Primary key checks:
https://github.com/vattle/sqlboiler/pull/138/files#diff-c880f21b86d224c850e56d2d0ce28501R226

If these are working we should be good to go!

**Comment by steffengy** [comment]
It lacks the equivalent of what the MySQL sqlboiler driver is doing:
 `if(extra = 'auto_increment','auto_increment', c.column_default),`.

Currently there isn't any check for autoincrement, I'll look into getting the equivalent (which is an ugly substring to check for the AUTOINCREMENT keyword in the SQL) in soon.

**Comment by steffengy** [comment]
@millerlogic Does this fix your issue with auto increment?

**Comment by millerlogic** [comment]
@steffengy it works if line 196 uses `?` instead of `%s`, thanks!

**Comment by steffengy** [comment]
Well OOPS. Still happy to hear that!

**Comment by maxmeyer** [comment]
Any progress on that PR?

**Comment by aarondl** [comment]
@maxmeyer The thing that's blocking it is the v3 way of doing drivers and the fact that merging it basically stops people from compiling `sqlboiler` on Windows unless they get the sqlite3 dll and mess around (same on Linux really, it just happens to be trivially easy on Linux).

We haven't been actively developing against v3 for some time now unfortunately but it's never left the backlog of things to do. We're about to kick off some new dev, and I think we might do it as a precursor to it.

**Comment by maxmeyer** [comment]
@aarondl That sounds good! Thanks for the update.

**Comment by aarondl** [comment]
Super happy to say that the work preceding this is finally done.

Here's an example of an external driver in the v3 branch:
https://github.com/volatiletech/sqlboiler/tree/v3/drivers/sqlboiler-psql

Because of the compilation trouble with this driver, I think it should remain outside the tree however. I will refactor and move it out if no one gets to it before me (this will be the last thing in v3 that I do, so there's plenty of time for someone to get to it before then!)

**Comment by aarondl** [comment]
There now exists an sqlboiler v3 driver for sqlite3: https://github.com/volatiletech/sqlboiler-sqlite3

I'll be announcing this with the release of v3 sometime very soon :)

**Comment by aarondl** [comment]
@steffengy If you desire we can add your name to copyrights in the license if you wish.
